plugins/logs/plugin.go

```diff
@@ -401,12 +401,46 @@ func (p *Plugin) Start(ctx context.Context) error {
 // Stop stops the plugin.
 func (p *Plugin) Stop(ctx context.Context) {
 	p.logInfo("Stopping decision logger.")
+
+	gracefulShutdownPeriod := p.manager.GracefulShutdownPeriod()
+	if p.config.Service != "" && gracefulShutdownPeriod > 0 && p.buffer.Len() > 0 {
+		p.flushDecisions(context.WithTimeout(ctx, time.Duration(gracefulShutdownPeriod)*time.Second))
+	}
+
 	done := make(chan struct{})
-	p.stop <- done
+	p.stop <-done
 	_ = <-done
 	p.manager.UpdatePluginStatus(Name, &plugins.Status{State: plugins.StateNotReady})
 }
 
+func (p *Plugin) flushDecisions(ctx context.Context, cancel context.CancelFunc) {
+	p.logInfo("Flushing decision logs.")
+	defer cancel()
+
+	go func(ctx context.Context, cancel context.CancelFunc) {
+		for {
+			ok, err := p.oneShot(ctx)
+			if err != nil {
+				p.logError("%v.", err)
+			} else if ok {
+				cancel()
+			}
+			// Wait some before retrying, but skip incrementing interval since we are shutting down
+			time.Sleep(1 * time.Second)
+		}
+	}(ctx, cancel)
+
+	select {
+	case <-ctx.Done():
+		switch ctx.Err() {
+		case context.DeadlineExceeded:
+			p.logInfo("Graceful shutdown period ended with decisions still in buffer.")
```

Comment:
Do we actually know that this is true? Could the spawned go routine have succeeded in draining the buffer in the meantime? (I guess the chances are very very small...)

Reply:
Hmm, I guess there's a small window in time where this could happen, however unlikely. Perhaps we could check `p.buffer.Len()` here? And only if there buffer is non-empty we print the current message (though probably at `logError`).

Reply:
Ah, no, the buffer length is not reliable for that since there could be messages logged without them having been put in the buffer yet. Changed the message to reflect this uncertainty, and changed log level to error.